edgelet/iotedged/src/lib.rs

```diff
@@ -430,35 +449,36 @@ where
 }
 
 #[cfg_attr(feature = "cargo-clippy", allow(too_many_arguments))]
-fn start_api<HC, K, F, C>(
+fn start_api<HC, K, F, C, W>(
     settings: &Settings<DockerConfig>,
     hyper_client: HC,
     runtime: &DockerModuleRuntime,
     key_store: &DerivedKeyStore<K>,
-    provisioning_result: &ProvisioningResult,
+    workload_config: W,
     root_key: K,
     shutdown_signal: F,
     crypto: &C,
     mut tokio_runtime: tokio::runtime::Runtime,
 ) -> Result<(), Error>
 where
     F: Future<Item = (), Error = ()> + Send + 'static,
-    HC: 'static + ClientImpl,
-    K: 'static + Sign + Clone + Send + Sync,
-    C: 'static
-        + CreateCertificate
+    HC: ClientImpl + 'static,
+    K: Sign + Clone + Send + Sync + 'static,
+    C: CreateCertificate
         + Decrypt
         + Encrypt
         + GetTrustBundle
         + MasterEncryptionKey
         + Clone
         + Send
-        + Sync,
+        + Sync
+        + 'static,
+    W: WorkloadConfig + Clone + Send + Sync + 'static,
 {
-    let hub_name = provisioning_result.hub_name();
-    let device_id = provisioning_result.device_id();
+    let hub_name = workload_config.iot_hub_name().to_string();
+    let device_id = workload_config.device_id().to_string();
     let hostname = format!("https://{}", hub_name);
-    let token_source = SasTokenSource::new(hub_name.to_string(), device_id.to_string(), root_key);
+    let token_source = SasTokenSource::new(hub_name.clone(), device_id.clone(), root_key);
```

Comment:
We can probably just say `workload_config.iot_hub_name().to_string()` and avoid an alloc. Same for device ID.

Reply:
SasTokenSource takes ownership of these strings and down below we use references to hub_name and device_id. Not sure we can do much here.